### PR TITLE
feat(alerts): add a regular job to detect anomalies

### PR DESCRIPTION
**Pull request by nikitaevg**
## Problem

https://github.com/PostHog/posthog/issues/14331

## Changes

This PR adds an initial version of the alerts notifications job. In the next PRs I'll introduce
1. Sending notifications if the alert calculation fails
2. Metrics for the number of alerts and number of anomalous alerts
3. UI warnings if users change the insight type to the ones that don't support alerts

## Does this work well for both Cloud and self-hosted?

Probably

## How did you test this code?

Automatic + manual testing

### CODE REVIEW

**Comment by nikitaevg** [comment]
Hi @mariusandra, PTAL!

File: frontend/src/lib/components/Alerts/views/EditAlert.tsx
Comment (mariusandra) [inline frontend/src/lib/components/Alerts/views/EditAlert.tsx:82]: nit:

```suggestion
                                    label="Upper threshold"
```

File: posthog/tasks/detect_alerts_anomalies.py
Comment (mariusandra) [reply posthog/tasks/detect_alerts_anomalies.py:28]: Yeah, this is correct :+1: . Currently we still have several insights floating around that only have filters (and no query), but the plan is to migrate everything over eventually.

**Review by mariusandra — COMMENTED** [review]
The logic seems reasonable to me, and since it's behind a flag I think we can merge it as is. 

However I do have some concerns about the longer term plan and would like to get a second opinion from @PostHog/team-product-analytics and also @benjackwhite (hog question below) and @pauldambra (loves dashboard reload cron jobs)

- Currently this will run once per hour at `x:20` and schedule a query to run for each alert immediately. Assuming we have 1000 alerts set up, this is a 1000 simultaneous queries every hour at the same time. We will need to stagger them somehow. For example cohort and dashboard reloads run more frequently, but then only run `n` oldest items every run leading to eventual "good enough" consistency. 
- The problem with dashboard and cohort calculations is nobody checks in. We periodically discover things have gotten worse when users complain. This will be worse if users will start to rely on alerts for their business. We'd need to establish _some_ practices around this, hence all the @ tagging above.
- Finally, we're hard at work on Hog and our CDP. It would be really cool to hook alerts into this system. @benjackwhite any thoughts on how to build the bridge?

File: posthog/tasks/detect_alerts_anomalies.py
Comment (benjackwhite) [inline posthog/tasks/detect_alerts_anomalies.py:58]: I think this is definitely not what we should do for a bunch of reasons:
1. No way to configure rate of delivery, backoffs, etc.
2. Email only is not the typical way people want to get alerted of this

We are building a new generic delivery system for the CDP (webhooks etc.) which would be the right place to have a destination and I think this _could_ play into that.

I don't want to pour water on the fire that is getting this work done as its super cool 😅 but I know that immediately we will have configuration and scaling issues here that I'm not sure we want to support.

I'm wondering if instead we could have an in-app only alert for now which we can then later hook up to the delivery service instead?

File: posthog/tasks/detect_alerts_anomalies.py
Comment (benjackwhite) [inline posthog/tasks/detect_alerts_anomalies.py:18]: I don't know for sure but this also feels like a scaling nightmare... We struggle sometimes to keep up with dashboard / insight refreshes in general and this is another form of refresh, just with a higher demand on reliability. I _think_ this would require strong co-ordination with @PostHog/team-product-analytics to make sure this fits in with their existing plans for improving background refreshing otherwise this will hit scaling issues fast.

**Review by benjackwhite — CHANGES_REQUESTED** [review]
Love the work here, but given how important and tricky this will be I'd like to consider a more minimal solution with input from @PostHog/team-product-analytics to make sure this is something we can actually scale.

**Review by nikitaevg — COMMENTED** [review]
Thanks for looking at it!

> Assuming we have 1000 alerts set up, this is a 1000 simultaneous queries every hour at the same time.

There's a way to set maximum parallel requests for celery, I think it should help to spread the load of this, no? 

> This will be worse if users will start to rely on alerts for their business. We'd need to establish some practices around this

I completely agree with that, it's not the final solution, just a skeleton. I'll need some help with this, but we need metrics and alerts about the job execution time to notice problems. I understand people will rely on alerts and it should be reliable alright.

File: posthog/tasks/detect_alerts_anomalies.py
Comment (webjunkie) [reply posthog/tasks/detect_alerts_anomalies.py:18]: > I think this would require strong co-ordination with @PostHog/team-product-analytics to make sure this fits in with their existing plans for improving background refreshing otherwise this will hit scaling issues fast.

Just wanted to chime in here. I can take a look at this, but am currently busy with being on support for this sprint. I'll see what we can do.

Comment (mariusandra) [reply posthog/tasks/detect_alerts_anomalies.py:18]: Scaling celery is not the issue, but ClickHouse will struggle and ultimately go down if suddenly 1000 simultaneous queries appear.

Comment (pauldambra) [reply posthog/tasks/detect_alerts_anomalies.py:18]: > should scale horizontally pretty easily - add a separate queue for these events, increase the number of parallel tasks in flight and add more servers if needed.

yep, was going to add "Should" is doing a lot of work in this sentence 😅 

@webjunkie I'm too far removed from how query code and caching interacts here

we already have one set of jobs that (is|should be) staying on top of having insight results readily available. does this use that cache? we should really overlap them so we have one set of tasks keeping a cache warm and then another that reads the fast access data in that cache for anomaly detection

humans aren't visiting insights once a minute so we know this will generate sustained load.

we should totally, totally build this feature - it's long overdue

--- 

i'm not opposed to getting a simple version in just for our team or select beta testers so we can validate the flow, but this 100% needs an internal sponsor since the work of rolling this out and scaling it simply can't be given to an external contributor (it wouldn't be fair or possible 🙈)

i would _love_ to be the internal sponsor but it's both not possible and completely outside of my current wheelhouse

(these concerns might be addressed elsewhere - i've not dug in here at all 🙈)

Comment (nikitaevg) [reply posthog/tasks/detect_alerts_anomalies.py:18]: > but ClickHouse will struggle and ultimately go down if suddenly 1000 simultaneous queries appear

Can't I limit the number of celery queries in flight? I understand this will introduce a problem of throughput, but then if the servers can't process N alerts each hour, maybe more read replicas or more servers are needed. I don't have much experience with column oriented databases though, so it's just a speculation.

> we already have one set of jobs that (is|should be) staying on top of having insight results readily available. does this use that cache?

🤷 , well the query_runner has some "cache" substrings in it's code, so one could assume... But I don't know

> humans aren't visiting insights once a minute so we know this will generate sustained load.

Just to clarify, it's once an hour

> but this 100% needs an internal sponsor since the work of rolling this out and scaling it simply can't be given to an external contributor (it wouldn't be fair or possible 🙈)

I completely agree and I would be really happy to have a mentor on this task. 

BTW, an interesting data point - Mixpanel [limits](https://docs.mixpanel.com/docs/features/alerts) their number of alerts to 50 per project.

Comment (webjunkie) [reply posthog/tasks/detect_alerts_anomalies.py:18]: We will talk among @PostHog/team-product-analytics next week and discuss this regarding ownership and so on.

Comment (pauldambra) [reply posthog/tasks/detect_alerts_anomalies.py:18]: > > humans aren't visiting insights once a minute so we know this will generate sustained load.
> Just to clarify, it's once an hour

👍 

(same point but thanks for clarification :))

File: posthog/models/alert.py
Comment (webjunkie) [inline posthog/models/alert.py:1]: I'd argue the model is too simplistic as is to support running alert checks at regular intervals.

I wrote up an RFC with what I think we should start with at https://github.com/PostHog/meta/pull/216

File: posthog/tasks/detect_alerts_anomalies.py
Comment (webjunkie) [inline posthog/tasks/detect_alerts_anomalies.py:27]: We should defer any query logic to existing methods, basically how it's done here

https://github.com/PostHog/posthog/blob/dd4be4fc3f97a85660db3e068c4ef002ddd5976b/posthog/api/insight.py#L575-L594

Execution mode should be `ExecutionMode.RECENT_CACHE_CALCULATE_BLOCKING_IF_STALE` so we leverage cache if available (and we have the benefit for the user of updating it here).

File: posthog/tasks/detect_alerts_anomalies.py
Comment (webjunkie) [inline posthog/tasks/detect_alerts_anomalies.py:25]: Please move task bodies into separate functions for testability and readability.

File: posthog/tasks/detect_alerts_anomalies.py
Comment (webjunkie) [inline posthog/tasks/detect_alerts_anomalies.py:54]: Notifying should be moved to separate functions for maintainability and for easier testing.

File: posthog/tasks/scheduled.py
Comment (webjunkie) [inline posthog/tasks/scheduled.py:251]: Should we start with each hour to begin with?

**Review by webjunkie — CHANGES_REQUESTED** [review]
Thanks for the contribution!

I think the general direction and scope of the PR is valid and something we can work with. Areas that needs to be improved before this can be merged is the workings of the Celery task and additional models and fields we need to sufficiently guide the execution.

I wrote up an RFC draft with how the Celery and model architecture could work:
https://github.com/PostHog/meta/pull/216

Let me know if this helps or needs discussion. (Either here or in Slack).

File: posthog/tasks/scheduled.py
Comment (pauldambra) [reply posthog/tasks/scheduled.py:251]: good call, tbh it should be configurable, someone self-hosting might want to set this to every second for example 🙈

File: posthog/tasks/detect_alerts_anomalies.py
Comment (nikitaevg) [reply posthog/tasks/detect_alerts_anomalies.py:30]: > for testability

I wouldn't say it hurts testability - one can call a task as a simple function

> and readability

I could argue that it doesn't help readability to have a function that just calls another function. Also, since we will have a separate task for sending notifications, we will have check_alert_task + check_alert + send_notification_task + send_notification. IMO it becomes cumbersome, I'll do it, but I'll be glad to revert it if you agree with me.

File: posthog/tasks/detect_alerts_anomalies.py
Comment (nikitaevg) [reply posthog/tasks/detect_alerts_anomalies.py]: Yeah, thanks, I was looking for something like this

File: posthog/tasks/detect_alerts_anomalies.py
Comment (nikitaevg) [reply posthog/tasks/detect_alerts_anomalies.py:67]: Sure, done

File: posthog/tasks/scheduled.py
Comment (nikitaevg) [reply posthog/tasks/scheduled.py:256]: Huh? Do you mean running the task at the beginning of each hour, not at 20 minutes?

File: ee/tasks/test/subscriptions/test_subscriptions.py
Comment (nikitaevg) [inline ee/tasks/test/subscriptions/test_subscriptions.py:27]: Just a redundant field

File: posthog/tasks/scheduled.py
Comment (webjunkie) [reply posthog/tasks/scheduled.py:256]: Sorry, read that line too quickly. We can leave it as is.

File: posthog/models/alert.py
Comment (webjunkie) [reply posthog/models/alert.py:1]: I agree with not creating too large PRs. However, I always find it useful to lay things out as needed later. As opposed to leaving a lot to refactor or clean up. Let's try to get the current scope of the PR merged.

File: posthog/tasks/alerts/test/test_checks.py
Comment (webjunkie) [inline posthog/tasks/alerts/test/test_checks.py:16]: You shouldn't need to patch or mock email directly. If you follow the recommendation of making email a sepearate task of at least function, you can easily mock that to see if notification has been called.

File: posthog/tasks/alerts/test/test_checks.py
Comment (webjunkie) [inline posthog/tasks/alerts/test/test_checks.py:23]: I'm not a fan of using API to test task behavior in task tests... as this couple theses tests to API implementation. Can we just use the model? (Or do we do this like that everywhere in PostHog 🙈?)

Comment (pauldambra) [reply posthog/tasks/alerts/test/test_checks.py:23]: there's more than one model involved in a dashboard, so this couples the tests to a single API regardless of what's beneath it... in this case the API since otherwise you can (and we did) break the API and not know about it because you're not using it

File: posthog/tasks/alerts/test/test_checks.py
Comment (webjunkie) [inline posthog/tasks/alerts/test/test_checks.py:22]: Ah, one other thing: Do not use `always_eager` in tests. The preferred way to test here is mocking. Mock the task and check that it is triggered. Test what the task does by testing what it calls (hence pulling out code out of the task into separate functions).

File: posthog/tasks/alerts/test/test_checks.py
Comment (nikitaevg) [reply posthog/tasks/alerts/test/test_checks.py:16]: Yeah, makes sense, will fix it soon

File: posthog/tasks/alerts/test/test_checks.py
Comment (nikitaevg) [reply posthog/tasks/alerts/test/test_checks.py:22]: I don't know, why not? I prefer tests without mocks over tests with mocks as long as the former is not very difficult to achieve. They cover more, less chances to miss a spot in the functionality.

File: posthog/tasks/alerts/test/test_checks.py
Comment (nikitaevg) [reply posthog/tasks/alerts/test/test_checks.py:23]: I prefer to test using public interfaces as it's usually much closer to what happens in production. Less importantly, it covers more functionality and usually simpler to write.

> as this couple theses tests to API implementation

I'm not sure it's a bad thing, see my point about much closer to production.

File: posthog/tasks/alerts/test/test_checks.py
Comment (webjunkie) [reply posthog/tasks/alerts/test/test_checks.py:22]: If you don't know and want to know why, please read the docs.

> The eager mode enabled by the [task_always_eager](https://celeryproject.readthedocs.io/zh-cn/latest/userguide/configuration.html#std:setting-task_always_eager) setting is by definition not suitable for unit tests.
>
> When testing with eager mode you are only testing an emulation of what happens in a worker, and there are many discrepancies between the emulation and what happens in reality.

Comment (nikitaevg) [reply posthog/tasks/alerts/test/test_checks.py:22]: Could you please add rationale to your change requests in the future? I couldn't know if it was your opinion or a documentation guideline, right? I was assuming the former.

Comment (webjunkie) [reply posthog/tasks/alerts/test/test_checks.py:22]: I see, sorry about that! I felt stating it as "preferred way" was enough to not make it seems as just my opinion.

Comment (nikitaevg) [reply posthog/tasks/alerts/test/test_checks.py]: Fixed

File: posthog/tasks/alerts/checks.py
Comment (webjunkie) [inline posthog/tasks/alerts/checks.py:80]: You can do `check_alert_task.si` (for immutable) above, then this doesn't happen/matter.

**Review by webjunkie — COMMENTED** [review]
This looks fine now considering the scope, but needs work in subsequent PRs as discussed.

File: posthog/tasks/alerts/checks.py
Comment (nikitaevg) [reply posthog/tasks/alerts/checks.py]: Yeah, it worked, thanks!

**Comment by nikitaevg** [comment]
@webjunkie Could you please merge this?

**Comment by nikitaevg** [comment]
@Twixes looks like Julian is unreachable, could you please merge this given the Julian's [approval](https://github.com/PostHog/posthog/pull/22762#pullrequestreview-2182433627)?

**Comment by nikitaevg** [comment]
@webjunkie, gentle reminder, could you please merge this?

**Review by webjunkie — APPROVED** [review]

File: frontend/src/types.ts
Comment (benjackwhite) [inline frontend/src/types.ts:4373]: TODO?

File: posthog/templates/email/alert_anomaly.html
Comment (benjackwhite) [inline posthog/templates/email/alert_anomaly.html:3]: Not blocking but "uh-oh" feels unnecessarily negative. The alert could be a positive thing.

File: posthog/tasks/alerts/checks.py
Comment (benjackwhite) [inline posthog/tasks/alerts/checks.py:83]: I don't think this needs to be a task. The `.send` function by default will queue a celery task for the actual sending.

**Review by benjackwhite — DISMISSED** [review]
A lot of hanging TODOs that I believe should be removed and a couple of smaller comments

File: frontend/src/types.ts
Comment (nikitaevg) [reply frontend/src/types.ts]: I was planning to do it in a follow up PR, but yeah, I can fix it here, done

File: posthog/tasks/alerts/checks.py
Comment (nikitaevg) [reply posthog/tasks/alerts/checks.py]: It makes sense, thanks, removed

File: posthog/templates/email/alert_anomaly.html
Comment (nikitaevg) [reply posthog/templates/email/alert_anomaly.html]: Removed :+1:

**Review by nikitaevg — COMMENTED** [review]
@benjackwhite PTAL!

**Comment by nikitaevg** [comment]
@webjunkie a gentle reminder, could you please take a look?

**Comment by nikitaevg** [comment]
@webjunkie do you know why all E2E tests might fail with the "Error: missing API token, please run `depot login`" error after merging with master?

**Comment by nikitaevg** [comment]
@benjackwhite could you please review this?

**Pull request merged**
